src/models/tree/build_tree.cu

```diff
@@ -180,75 +181,356 @@ __global__ static void __launch_bounds__(BLOCK_THREADS)
-                      int64_t seed)
-{
-  constexpr int32_t WarpSize = 32;
-  const int32_t warp_id      = threadIdx.x / WarpSize;
-  const int32_t lane_id      = threadIdx.x % WarpSize;
-
-  const int32_t localIdx = blockIdx.x * TPB + warp_id * WarpSize + lane_id;
-
-  // prefetch sampleNode information for all 32 ids
-  auto [sampleNode_lane, localSampleId_lane] = (localIdx < batch.InstancesInBatch())
-                                                 ? batch.instances_begin[localIdx]
-                                                 : cuda::std::make_tuple(-1, -1);
-  const bool computeHistogram =
-    localIdx < batch.InstancesInBatch() &&
-    ComputeHistogramBin(
-      sampleNode_lane, node_sums, histogram.ContainsNode(BinaryTree::Parent(sampleNode_lane)));
-
-  // mask contains all sample bits of the next 32 ids that need to be bin'ed
-  auto lane_mask = ballot(computeHistogram);
-
-  // reverse to use __clz instead of __ffs
-  lane_mask = __brev(lane_mask);
-
-  while (lane_mask) {
-    // look for next lane_offset / sample to process within warp-batch
-    const uint32_t lane_offset  = __clz(lane_mask);
-    const int32_t sampleNode    = shfl(sampleNode_lane, lane_offset);
-    const int32_t localSampleId = shfl(localSampleId_lane, lane_offset);
-
-    // remove lane_offset bit from lane_mask for next iteration
-    lane_mask &= (0x7fffffff >> lane_offset);
-
-    auto feature_begin = blockIdx.y * FEATURES_PER_WARP;
-    auto feature_end   = min(n_features, (size_t)feature_begin + FEATURES_PER_WARP);
-    for (int32_t feature = feature_begin + lane_id; feature < feature_end; feature += WarpSize) {
-      const int32_t bin_idx =
-        split_proposals.FindBin(X[{sample_offset + localSampleId, feature, 0}], feature);
-      for (int32_t output = 0; output < n_outputs; output++) {
-        // get same G/H from every thread in warp
-        legate::Point<3> p = {sample_offset + localSampleId, feature, output};
-        auto gpair_quantised =
-          quantiser.QuantiseStochasticRounding({g[p], h[p]}, hash_combine(seed, p[0], p[2]));
-        auto* addPosition = reinterpret_cast<typename IntegerGPair::value_type*>(
-          &histogram[{sampleNode, output, bin_idx}]);
-
-        if (bin_idx != SparseSplitProposals<TYPE>::NOT_FOUND) {
-          Histogram<IntegerGPair>::atomic_add_type* addPosition =
-            reinterpret_cast<Histogram<IntegerGPair>::atomic_add_type*>(
-              &histogram[{sampleNode, output, bin_idx}]);
-          atomicAdd(addPosition, gpair_quantised.grad);
-          atomicAdd(addPosition + 1, gpair_quantised.hess);
-        }
-      }
-    }
+                      int64_t seed,
+                      cudaStream_t stream)
+  {
+    int average_features_per_group         = split_proposals.num_features / num_groups;
+    std::size_t average_elements_per_group = batch.InstancesInBatch() * average_features_per_group;
+    auto min_blocks  = (average_elements_per_group + kItemsPerTile - 1) / kItemsPerTile;
```

Comment:
Yes the average features is not a great heuristic. This code is adapted from https://github.com/dmlc/xgboost/blob/18edf86b0acad4db3ed350d806f127da65d46fec/src/tree/gpu_hist/histogram.cu#L377.

Its an attempt to not launch a too large grid only when the amount of data is small. This is probably more relevant to xgboost which processes one node at a time, where we are more likely to be processing a larger batch. It could also be relevant to small datasets.